### ISSUE TITLE
Missing patch for using shiftfs in conjunction with overlayfs

### ISSUE THREAD

**Issue by hxx-fetch** [toby63/shiftfs-dkms#2]
There seems to be an additional kernel patch in conjunction with shiftfs when used above overlayfs:

https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/focal/commit/fs/overlayfs?id=d24b8a547be1578cb5a200ad9ac57a258f0a9de1

So it should be noted in the documentation not to use this dkms in conjunction with overlayfs.

**Comment by toby63** [comment]
@hxx-fetch: 
You are right. 
Although this problem is mentioned in a post I linked to in the Readme (see "Further information on shiftfs"), I agree that I should clarify it here in the Readme too.  
[Quote](https://discuss.linuxcontainers.org/t/trying-out-shiftfs/5155) from the LXD Forum:
> Limitations     
> 
> The only limitation we’re currently aware of is that shiftfs will prevent the use of overlayfs inside the container. This may break Docker users. If this applies to you, you should stay away from shiftfs for now.   

I added a note in the Readme ([see commit](https://github.com/toby63/shiftfs-dkms/commit/9d5816789dec00563090e30098406164c526ea2b)).

**Comment by toby63** [comment]
I was not aware of a patch for this though, interesting :thinking: .

@brauner: Some questions about this:
1. Just for clarity: This problem only occurs inside containers?
So using overlayfs on the host still works (e.g. using docker on the host)?

2. Will the [patch(es)](https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/focal/commit/fs/overlayfs?id=d24b8a547be1578cb5a200ad9ac57a258f0a9de1) be in the mainline kernel?
And what is the status of implementation?

3. I assume it is not possible or "easy" to put the patch(es) into a kernel module?

**Comment by brauner** [comment]
> I was not aware of a patch for this though, interesting  .
> 
> @brauner: Some questions about this:
> 
> 1. Just for clarity: This problem only occurs inside containers?
>    So using overlayfs on the host still works (e.g. using docker on the host)?

Yes.

> 2. Will the [patch(es)](https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/focal/commit/fs/overlayfs?id=d24b8a547be1578cb5a200ad9ac57a258f0a9de1) be in the mainline kernel?
>    If yes, what is the status of implementation?

We're working on an upstreamable version of shiftfs but it'll be a little while.

> 3. I assume it is not possible or "easy" to put the patch(es) into a kernel module?

Not impossible but unlikely to be accepted.

**Comment by toby63** [comment]
@brauner: Thank you the information.

> > 3. I assume it is not possible or "easy" to put the patch(es) into a kernel module?
> 
> Not impossible but unlikely to be accepted.

Sry for asking, but that means:
a) accepted by the kernel maintainers?
b) accepted by the system/kernel (for example my system)?

And it would require a rewrite I assume?

I was just thinking of implementing that into a dkms skript once again :smile: .

**Comment by brauner** [comment]
> @brauner: Thank you for the information.
> 
> > > 1. I assume it is not possible or "easy" to put the patch(es) into a kernel module?
> > 
> > 
> > Not impossible but unlikely to be accepted.
> 
> Sry for asking, but that means:
> a) accepted by the kernel maintainers?

Yes, unlikely to be accepted upstream because we're (as in us, the kernel maintainers) how exactly we want this feature to look upstream.

> b) accepted by the system/kernel (for example my system)?

No, I meant a). I don't know what your maintainers would say. :)

> 
> And it would require a rewrite I assume?
> 
> I was just thinking of implementing that into a dkms skript once again  .

If you can find a way for that patch to be included in the dkms script, then sure. :)

**Comment by toby63** [comment]
> If you can find a way for that patch to be included in the dkms script, then sure. :)

Really? I thought that for example shiftfs was only so "easily" implementable into a kernel module (I mean for me in this skript) because it was limited in some way ("just" an additional file that did not exist before in the kernel).
In this case your patches seem to alter different files that already exist. Isn't that a problem?
So kernel modules (or their installation) can alter other files in the kernel? 
Additional question: Isn't that prohibited (I mean like some "unofficial rule")? 
Or do you mean that I have to find a workaround like the magic number (probably bad example) that I included in the Makefile (EXTRA_CFLAGS := -DSHIFTFS_MAGIC=0x6a656a62), instead of writing it into the existing magic file?
Sry I know nearly nothing about kernel work :smile: , but I want to ask in case someone (maybe even me) wants to try this.

**Comment by toby63** [comment]
Found something:
https://www.reddit.com/r/linux/comments/6g4l7n/using_dkms_to_compile_patched_modules/

I will probably test this in a few days.

**Comment by toby63** [comment]
@brauner: 
I also found a problem:

Your [commit](https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/focal/commit/fs/overlayfs?id=d24b8a547be1578cb5a200ad9ac57a258f0a9de1) includes a check for **SHIFTFS_MAGIC**:
```
#fs/overlayfs/super.c:
if (!err) {
			if ((path->dentry->d_sb->s_magic != SHIFTFS_MAGIC) && ovl_dentry_remote(path->dentry)) {
```

In my dkms script for shiftfs I did not add the magic number to "include/uapi/linux/magic.h" (maybe not possible with dkms?).
Instead I used a workaround [suggested by stgraber](https://discuss.linuxcontainers.org/t/trying-out-shiftfs/5155/41): 
I only added the magic number to the Makefile of shiftfs (via EXTRA_CFLAGS).

So I guess overlayfs cannot read the magic number.
I would solve this by also adding the magic number via EXTRA_CFLAGS to the Makefile of overlayfs, or would you suggest a different solution?

**Comment by toby63** [comment]
@hxx-fetch Sry to bother, 
am I assuming right that you use the docker setup this problem is about?
(just as a reminder, use of docker with overlayfs inside lxd container )   
If yes; I am testing the docker setup right now, and somehow it does not work. 
Now I want to rule out certain setup mistakes.

So it would be nice if you could answer me some questions:
* what storage backend do you use for lxd?
* what image (or OS) do you use for the lxd container?
* What version of docker do you use (inside the container of course)?
* Do you use the overlay2 storage driver in docker?
* did you use any other special options for lxd or docker that could be relevant for this setup?

**Comment by hxx-fetch** [comment]
@toby63 
Hi,

Sorry, I don't think I'm much of help with this issue:

I was looking for a solution how to get shiftfs run on gentoo-linux,
not wanting to manage custom kernel patches by myself. I then found 
your project and seeing the overlayfs-patches were not included,
I thought I give you a notification so docker/overlayfs users won't have 
unexpected problems.

So what I have currently running is lxd 3.16 on gentoo-linux using zfs, but 
no docker/overlayfs. The containers are debian, ubuntu and gentoo
I plan to use your shiftfs dkms when lxd 4.0 will be avail in gentoo.

**Comment by toby63** [comment]
@hxx-fetch 
Thank you for answering anyway :slightly_smiling_face: .
Also glad my project is useful for you & thank you for the issue report, because I was aware of this problem, but I haven't mentioned it here and I also was not aware of available patches.

In general (for everyone reading this):
I have one idea left to test, but if that does not run, I am out of clues :smile: .
But honestly this problem seems to be veeery special to me more and more.
Because the standard storage backends for lxd (btrfs or zfs (at least thats the recommendation)) and docker (overlayfs) are not compatible (at least thats my impression).   
So docker will either use btrfs or zfs as well or you need to use a different backend for lxd (dir with ext4 on hosts partition; which is what I try to test).
